### ISSUE TITLE
Jetty considers bootstrap injected class to be a "server class"

### ISSUE THREAD

**Issue by cataphract** [eclipse/jetty.project#4183]
Version: 9.4.20.v20190813

If a servlet constructor depends on a bootstrap class (a class injected into the bootstrap classloader) whose location cannot be found, jetty will consider it a "server class" and `WebAppClassLoader` will refuse to return it.

`_context.isServerClass(parentClass)` returns true here:

```java
                    parentClass = _parent.loadClass(name);
                    // If the webapp is allowed to see this class
                    if (Boolean.TRUE.equals(__loadServerClasses.get()) || !_context.isServerClass(parentClass))
                    {
                        if (LOG.isDebugEnabled())
                            LOG.debug("WAP parent loaded {}", parentClass);
                        return parentClass;
                    }
```

In `ClasspathPattern`:

```java
    public boolean match(Class<?> clazz)
    {
        try
        {
            return combine(_packageOrNamePatterns, clazz.getName(), _locations, () -> TypeUtil.getLocationOfClass(clazz));
        }
        catch (Exception e)
        {
            LOG.warn(e);
        }
        return false;
    }
```

`TypeUtil.getLocationOfClass(java.util.List.class)` is `file:/opt/java/openjdk/lib/rt.jar`, `TypeUtil.getLocationOfClass(java.io.sqreen.agent.instrumentation.boot.BlockedBySqreenAdviceException.class)` is `null` (because its classloader and protection domain are null, so none of the two methods for finding the location succeeds).

```java
    private static boolean combine(IncludeExcludeSet<Entry, String> names, String name, IncludeExcludeSet<Entry, URI> locations, Supplier<URI> location)
    {
        Boolean byName = names.isIncludedAndNotExcluded(name);
        if (Boolean.FALSE == byName)
            return false;

        URI uri = location.get();
        if (uri == null)
            return locations.isEmpty() || locations.hasExcludes() && !locations.hasIncludes();
```

`location.get()` is `null`, and `locations` is empty, so this returns true.

End result:

<pre>
Suppressed: java.lang.NoClassDefFoundError: java/io/sqreen/agent/instrumentation/boot/BlockedBySqreenAdviceException
at java.lang.Class.getDeclaredConstructors0(Native Method)
at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
at java.lang.Class.getConstructor0(Class.java:3075)
at java.lang.Class.getDeclaredConstructor(Class.java:2178)
at org.eclipse.jetty.server.handler.ContextHandler$Context.createInstance(ContextHandler.java:2531)
at org.eclipse.jetty.servlet.ServletContextHandler$Context.createServlet(ServletContextHandler.java:1291)
at org.eclipse.jetty.servlet.ServletHolder.newInstance(ServletHolder.java:1270)
at org.eclipse.jetty.servlet.ServletHolder.initServlet(ServletHolder.java:633)
at org.eclipse.jetty.servlet.ServletHolder.initialize(ServletHolder.java:421)
at org.eclipse.jetty.servlet.ServletHandler.lambda$initialize$0(ServletHandler.java:746)
at java.util.stream.SortedOps$SizedRefSortingSink.end(SortedOps.java:357)
</pre>

**Comment by janbartel** [comment]
It would be good if you could provide a reproduction test case. 

As a workaround, have you tried specifying the System and Server classes so that your class is considered a System class, and not a Server class? You can use WebAppContext.getSystemClasspathPattern().add("java.io.sqreen.agent.instrumentation.boot.") and WebAppContext.getServerClasspathPattern().add("-java.io.sqreen.agent.instrumentation.boot.")

**Comment by cataphract** [comment]
A minimal test case would involve writing an agent that injects a class into the bootstrap classloader and then instruments code that's loaded by the webappclassloader depending on such injected class. The class name in question already starts with `java.` precisely to avoid the sort of filters some application servers have.

In any case, to reproduce you can use this Dockerfile:
```dockerfile
FROM owasp/dependency-track

USER root

RUN apk add --update \
    curl \
    && rm -rf /var/cache/apk/*
RUN curl https://s3-eu-west-1.amazonaws.com/sqreen-download-public/java/sqreen-1.6.1.jar -o sqreen.jar

CMD java -javaagent:sqreen.jar -Dsqreen.app_name=java-deptrack $JAVA_OPTIONS -DdependencyTrack.logging.level=$LOGGING_LEVEL -jar dependency-track-embedded.war
```

The exception can be seen with `docker exec <container> cat /root/.dependency-track/server.log`

The workaround I ended up with was instrumenting `isServerClass(java.lang.Class)` of classes implementing `org.eclipse.jetty.util.ClassVisibilityChecker` to let my classes through.

**Comment by janbartel** [comment]
Created PR #4197.

**Comment by janbartel** [comment]
Issue resolved via PR #4197
